## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [104]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [105]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [106]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [107]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [108]:
df.select_dtypes(include=['object']).head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  order_date         item_id  \
0     harding.donald.7185@sullivan.com  2023-04-28  EUD29711-63-6U   
1    myers.t.6537@ferguson-johnson.net  2023-05-19  XDA18116-89-4A   
2                swatson8146@payne.net  2023-01-29  ABE59463-05-7E   
3               bakernathan@benson.com  2023-04-25  ZMM00836-65-0C   
4  christinaschwartz9252@mcconnell.com  2023-05-28  BZX55559-12-3X   

      category         subcategory  
0        decor            wall art  
1  consumables                pens  
2     software  project management  
3  consumables                pens  
4  consumables                misc

In [109]:
df.values

array([['Donald', 'Harding', 'Immunologist', ..., 7.5, 105, 1],
       ['Tiffany', 'Myers', 'Music therapist', ..., 1.49, 21, 0],
       ['Shannon', 'Watson', 'Immunologist', ..., 1.68, 39, 6],
       ...,
       ['Monica', 'Gutierrez', 'Graphic designer', ..., 4.16, 475, 6],
       ['Wanda', 'Solomon', 'Toxicologist', ..., 18.04, 112, 0],
       ['Dwayne', 'Bell', 'Chief Financial Officer', ..., 2.07, 1031, 5]],
      dtype=object)

In [110]:
df.axes

[RangeIndex(start=0, stop=54639, step=1),
 Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
        'order_date', 'order_week', 'order_year', 'item_id', 'category',
        'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
        'line_number'],
       dtype='object')]

In [111]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [112]:
# For the category with the most entries, which subcategory had the most entries?
df[df['category'] == 'consumables']['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [113]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head()

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [114]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head().index.tolist()
top_5_clients

[33615, 66037, 46820, 38378, 24741]

In [115]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_total_qty = df.loc[df['client_id'] == top_5_clients[0],'qty']
top_client_total_qty.sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [116]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df[['unit_price','qty','subtotal']].head()


unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00

In [117]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df.apply(lambda x: x['total_weight'] * 7 if x['total_weight'] > 50 else x['total_weight'] * 10, axis=1)
df[['unit_price','unit_weight','qty','total_weight', 'shipping_price']].head()

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [118]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price'] = df['subtotal'] + df['shipping_price'] + (df['subtotal'] + df['shipping_price']) * 0.0925
df['line_price'] = df['line_price'].round(2)
df[['subtotal','shipping_price','line_price']].head(3)

subtotal  shipping_price  line_price
0  115164.00         5512.50   131839.08
1     523.95          312.90      914.26
2     527.28          458.64     1077.12

In [119]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_price']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  total_weight  shipping_price  line_price  line_cost  
0            1  115164.00        787.50         5512.50   131839.08   85597.05  
1            0     523.95         31.29          312.90      914.26     629.79  
2            6     527.28         65.52          458.64     1077.12     765.18  
3            3    1056.18         35.67          356.70     1543.57    1077.35  
4            1    3902.00        928.60         6500.20    11364.40    8663.60  

[5 rows x 23 columns]

In [120]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    subtotal  total_weight  shipping_price  line_price  line_cost  line_profit  
0  115164.00        787.50         5512.50   131839.08   85597.05     46242.03  
1     523.95         31.29          312.90      914.26     629.79       284.47  
2     527.28         65.52          458.64     1077.12     765.18       311.94  
3    1056.18         35.67          356.70     1543.57    1077.35       466.22  
4    3902.00        928.60         6500.20    11364.40    8663.60      2700.80  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [121]:
# Check your work using the totals above
order_2742071 = df.loc[df['order_id'] == 2742071, 'line_price']
order_2173913 = df.loc[df['order_id'] == 2173913, 'line_price']
order_6128929 = df.loc[df['order_id'] == 6128929, 'line_price']
print(f"Order ID 2742071 had a total price of ${order_2742071.sum():,.2f}")
print(f"Order ID 2173913 had a total price of ${order_2173913.sum():,.2f}")
print(f"Order ID 6128929 had a total price of ${order_6128929.sum():,.2f}")

Order ID 2742071 had a total price of $152,811.89
Order ID 2173913 had a total price of $162,388.71
Order ID 6128929 had a total price of $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [122]:
df.loc[df['client_id'] == 33615, 'qty'].sum()

64313

In [123]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in top_5_clients:
    total_spent = df.loc[df['client_id'] == client, 'line_price'].sum()
    print(f"{client}: ${total_spent:.2f}")


33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [124]:
# What I thought the previous code block was asking for (got the top 5 clients by quantity, calculated the total spent by each, then sorted by their total spent in descending order)
clients = df['client_id'].unique()
top_clients_total = []
for client in clients:
    qty_total = df.loc[df['client_id'] == client, 'qty'].sum()
    total_spent = df.loc[df['client_id'] == client, 'line_price'].sum()
    top_clients_total.append({'client_id': client, 'qty': qty_total, 'spent': total_spent})
top_clients_total = pd.DataFrame(top_clients_total)
top_clients_total = top_clients_total.sort_values(by='qty', ascending=False)
top_clients_total.head()

client_id      qty         spent
267      78965  4033302  6.559237e+08
435      86721  1010426  1.812663e+08
386      30143   846069  3.332885e+07
389      85931   644423  2.374170e+08
774      77354   506382  7.194265e+07

In [125]:
top_5_clients_summary = []
for client in top_5_clients:
    top_5_clients_summary.append(df.loc[df['client_id'] == client].sum())
top_5_clients_summary = pd.DataFrame(top_5_clients_summary)
top_5_clients_summary[['client_id', 'qty']].head()

client_id     qty
0    7395300   64313
1   13933807   43018
2    9785380   75768
3    7944246   73667
4    5121387  239862

In [126]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5_clients_summary = df.groupby('client_id')[['qty','shipping_price','line_price','line_cost','line_profit']].sum()
top_5_clients_summary = top_5_clients_summary.loc[top_5_clients]
top_5_clients_summary = top_5_clients_summary.reset_index()
top_5_clients_summary.head()

client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [127]:
# Format the data and rename the columns to names suitable for presentation.
top_5_clients_summary.rename(columns={'client_id': 'Client ID', 'qty': 'Units', 'shipping_price': 'Shipping', 'line_price': 'Total Revenue', 'line_cost': 'Total Cost', 'line_profit': 'Total Profit'}, inplace=True)

# Define the money columns. 
money_columns = ['Shipping', 'Total Revenue', 'Total Cost', 'Total Profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
        return value/1000000

# Apply the currency_format_millions function to only the money columns. 
for column in money_columns:
    top_5_clients_summary[column] = top_5_clients_summary[column].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
top_5_clients_summary.rename(columns={'Shipping': 'Shipping (millions)', 'Total Revenue': 'Total Revenue (millions)', 'Total Cost': 'Total Cost (millions)', 'Total Profit': 'Total Profit (millions)'}, inplace=True)
top_5_clients_summary



Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [128]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
top_5_clients_summary_sorted = top_5_clients_summary.sort_values(by='Total Profit (millions)', ascending=False)
top_5_clients_summary_sorted

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995